system-probe: Add support for building and testing Rust binary

vitkyrka opened the pull request:
### What does this PR do?

As a first step towards integrating the minimal Rust privileged agent for
discovery, add basic support for building a Rust binary and invoking it from
the go tests, including in KMT. The binary is not shipped.

Since this is the first use of the Rust toolchain in the Datadog agent build,
this is done using a dummy program in order to separate any build, toolchain
and CI related issues and concerns from any concerns about the actual new
privileged agent (which will be imported in a separate, later step).

Note that this also does not deal with the question of invoking Rust unit tests
in the agent build and in KMT, that will also be dealt with separately.

### Motivation

https://datadoghq.atlassian.net/browse/DSCVR-313

### Describe how you validated your changes

CI

### Additional Notes

The Rust binary is not built statically and due to this has issues running on
Centos 7.9 which likely has an older glibc (I wasn't able to debug this fully
since I had issues setting up that distro in KMT locally).

Building the binary statically with -C target-feature=+crt-static did not work
due to rcrt1.o apparently missing in the toolchain.

A fix could have been to include and use the musl toolchain which is the one
recommended for statically compliing Rust, but instead of complicating things
further I just skipped the tests on Centos 7.x for now.

vitkyrka committed:
rust

usamasaqib reviewed (approved):
Looks good to me. Just added a suggestion for resolving the glibc issue.

vitkyrka committed:
Set executable permissions in Bazel

vitkyrka committed:
Skip on macOS

vitkyrka committed:
Buildifier

vitkyrka commented on tasks/system_probe.py:
```diff
@@ -1614,6 +1620,28 @@ def cp_cmd(out_dir):
+    platform_map = {
+        "x86_64": "//bazel/platforms:linux_x86_64",
+        "arm64": "//bazel/platforms:linux_arm64",
+    }
```
This code path is not called on macOS, but I've added a check to skip it there explicitly now just like the Windows check.

rdesgroppes reviewed (approved):
LGTM

vitkyrka commented on pkg/collector/corechecks/servicediscovery/module/impl_linux_test.go:
```diff
@@ -418,3 +418,29 @@ ffffb7360000-ffffb74ec000 r-xp 00000000 00:22 13920                      /opt/ot
+	platformVersion, err := kernel.PlatformVersion()
+	require.NoError(t, err)
+
+	if platform == "centos" && strings.HasPrefix(platformVersion, "7") {
```
I think one part of the problem is that the C/C++ toolchain we use for Bazel [doesn't support](https://github.com/rust-lang/rust/issues/99382#issuecomment-1188871914) `-static-pie`, otherwise we could try to build the test binaries statically. That toolchain is built by us (in some other repo) so we could maybe change it to be built with the appropriate support.

```
vagrant@agent-dev-ubuntu-24:~$ echo 'int main(){}' | gcc -static-pie -x c -
[0.067s] vagrant@agent-dev-ubuntu-24:~$ echo 'int main(){}' | .cache/bazel/_bazel_vagrant/8a1be3ba35926e170a8cb57e37b5b3bd/external/gcc_toolchain++gcc_toolchains+gcc_toolchain_aarch64/xbin/gcc -static-pie -x c -
/home/vagrant/.cache/bazel/_bazel_vagrant/8a1be3ba35926e170a8cb57e37b5b3bd/external/gcc_toolchain++gcc_toolchains+gcc_toolchain_aarch64/bin/../lib/gcc/aarch64-unknown-linux-gnu/12.3.0/../../../../aarch64-unknown-linux-gnu/bin/ld.bfd: cannot find rcrt1.o: No such file or directory
collect2: error: ld returned 1 exit status
[0.046s] 1  vagrant@agent-dev-ubuntu-24:~$ echo 'int main(){}' | .cache/bazel/_bazel_vagrant/8a1be3ba35926e170a8cb57e37b5b3bd/external/gcc_toolchain++gcc_toolchains+gcc_toolchain_aarch64/xbin/gcc -static -x c -
[0.081s] vagrant@agent-dev-ubuntu-24:~$
```

vitkyrka committed:
bazel: Enable rustfmt and clippy in CI (#44830)

In CI, enable rustfmt and clippy checks for Rust targets during the
build phase.

This ensures that the rules are enforced by CI while still allowing
developers
to temporarily ignore them locally during development.

vitkyrka committed:
Merge remote-tracking branch 'origin/main' into vitkyrka/rust

 Conflicts:
	MODULE.bazel
	MODULE.bazel.lock

vitkyrka commented:
/merge
